### PR TITLE
Disposable - Replace magazines in backpack when launcher fits backpacks

### PR DESCRIPTION
**Pull request by 10Dozen**
**When merged this pull request will:**
- Allow replacement of magazines in backpacks with launchers, if loaded launcher item fits backpacks (has backpack slot id in `launcher_loaded >> WeaponSlotsInfo >> allowedSlots` array, implements https://github.com/CBATeam/CBA_A3/issues/1356)

### CODE REVIEW

File: addons/disposable/XEH_preInit.sqf [hunk 1]
```diff
@@ -27,6 +27,7 @@ GVAR(NormalLaunchers) = [] call CBA_fnc_createNamespace;
 GVAR(LoadedLaunchers) = [] call CBA_fnc_createNamespace;
 GVAR(UsedLaunchers) = [] call CBA_fnc_createNamespace;
 GVAR(magazines) = [];
+GVAR(BackpackReplaceableMagazines) = [];
```
Comment (commy2) [inline addons/disposable/XEH_preInit.sqf:30]:
```suggestion
GVAR(BackpackLaunchers) = createHashMap;
```

File: addons/disposable/XEH_preInit.sqf [hunk 2]
```diff
@@ -35,6 +36,7 @@ private _cfgMagazines = configFile >> "CfgMagazines";
 {
     private _launcher = configName _x;
     private _magazine = configName (_cfgMagazines >> (getArray (_cfgWeapons >> _launcher >> "magazines") select 0));
+    private _replaceInBackpack = [_cfgMagazines >> _magazine, QGVAR(launcherFitsBackpack), 0] call BIS_fnc_returnConfigEntry > 0;
```
Comment (commy2) [inline addons/disposable/XEH_preInit.sqf:39]:
```suggestion
    private _fitsInBackpacks = TYPE_BACKPACK in getArray(configFile >> "CfgWeapons" >> _launcher >> "WeaponSlotsInfo" >> "allowedSlots");
```
with
```sqf
#define TYPE_VEST 701
#define TYPE_UNIFORM 801
#define TYPE_BACKPACK 901
```
somewhere on top or macro header script_components.hpp file.

File: addons/disposable/XEH_preInit.sqf [hunk 3]
```diff
@@ -48,6 +50,10 @@ private _cfgMagazines = configFile >> "CfgMagazines";
         GVAR(MagazineLaunchers) setVariable [_magazine, _loadedLauncher];
     };
 
+    if (_replaceInBackpack) then {
+        GVAR(BackpackReplaceableMagazines) pushBackUnique _magazine;
```
Comment (commy2) [inline addons/disposable/XEH_preInit.sqf:54]:
```suggestion
    if (_fitsInBackpacks) then {
        GVAR(BackpackLaunchers) set [_launcher, true];
```

File: addons/disposable/fnc_replaceMagazineCargo.sqf [hunk 4]
```diff
@@ -56,7 +56,7 @@ private _isBackpack = getNumber (configOf _box >> "isBackpack") != -1;
     _x params ["_magazine", "_ammo"];
 
     if (_magazine in GVAR(magazines)) then {
-        if !(_isBackpack) then {
+        if (!_isBackpack || {_magazine in GVAR(BackpackReplaceableMagazines)}) then {
             _box addWeaponCargoGlobal [GVAR(MagazineLaunchers) getVariable _magazine, 1];
```
Comment (commy2) [inline addons/disposable/fnc_replaceMagazineCargo.sqf:60]:
```suggestion
        private _launcher = GVAR(MagazineLaunchers) getVariable _magazine;
        if (!_isBackpack || _launcher in GVAR(BackpackLaunchers)) then {
            _box addWeaponCargoGlobal [_launcher, 1];
```

**Review by commy2 — CHANGES_REQUESTED** [review]
1. No need to make a config entry for this, because it already is one with the "WeaponsSlotsInfo\allowedSlots" array.
2. No need to maintain a list when we can abuse HashMaps as sets. Lookup in list is O(n) at worts, while lookup in set is O(1).

Disclaimer, suggestions not tested, so there may be typos or logic holes.

**Comment by 10Dozen** [comment]
> commy2 is like: I'll never add this feature! 
Also commy2: I rewrite every single line of the feature's PR!

`>_<`

File: addons/disposable/XEH_preInit.sqf (hunk 2 above)
Comment (commy2) [reply addons/disposable/XEH_preInit.sqf:39]: May need to be moved down and _launcher be `_loadedLauncher`.

**Comment by commy2** [comment]
Such is life.

File: addons/disposable/XEH_preInit.sqf (hunk 1 above)
Comment (10Dozen) [reply addons/disposable/XEH_preInit.sqf:30]: Should i ~~break~~ change other lists/namespaces into hashMaps too?

**Comment by commy2** [comment]
lgtm

Did you test if it works?

**Comment by 10Dozen** [comment]
E.g. for CUP, as i understand, some ammo was intentionally made lighter/smaller than launcher to make ATs more vanilla-like. But when converted to launchers - disposals takes too much space. 
I feel like some extra space check is needed on conversion in that case (e.g. PzF3 + 3 rockets converts to PzF3 + 1 extra PzF3 in backpack, the rest is ignored as there is no more space).

**Comment by 10Dozen** [comment]
> Did you test if it works?

Yeah, works fine.

**Review by commy2 — APPROVED** [review]

**Comment by commy2** [comment]
> Should i break change other lists/namespaces into hashMaps too?

I don't know. It seems not really necessary.

File: addons/disposable/XEH_preInit.sqf [hunk 5]
```diff
@@ -36,6 +37,7 @@ private _cfgMagazines = configFile >> "CfgMagazines";
     private _launcher = configName _x;
     private _magazine = configName (_cfgMagazines >> (getArray (_cfgWeapons >> _launcher >> "magazines") select 0));
     getArray _x params ["_loadedLauncher", "_usedLauncher"];
+    private _fitsInBackpacks = TYPE_BACKPACK in getArray(configFile >> "CfgWeapons" >> _loadedLauncher >> "WeaponSlotsInfo" >> "allowedSlots");
```
Comment (commy2) [inline addons/disposable/XEH_preInit.sqf:40]:
```suggestion
     private _fitsInBackpacks = TYPE_BACKPACK in getArray (configFile >> "CfgWeapons" >> _loadedLauncher >> "WeaponSlotsInfo" >> "allowedSlots");
```
We need a PEP-8 of SQF.

**Comment by commy2** [comment]
Good job :)

**Pull request merged**